File: leetcode/middle/207/207.py
from typing import List
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indeg = [0 for _ in range(numCourses)]
        mp = [[] for _ in range(numCourses)]
        
        for i in prerequisites:
            indeg[i[0]] += 1
            mp[i[1]].append(i[0])
        q = []
        tot = 0
        for i in range(numCourses):
            if indeg[i] == 0:
                q.append(i)
                tot+=1

        while q:
            x = q.pop(0)
            for i in mp[x]:
                indeg[i] -= 1
                if indeg[i] == 0:
                    q.append(i)
                    tot += 1
        return tot == numCourses


if __name__ == '__main__':
    ret = Solution().canFinish(2, [[1,0]])
    print(ret)

File: leetcode/middle/207/207.cpp
#include<iostream>
#include<vector>
#include<set>
#include<string>
#include<queue>
#include<cstring>
using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) 
    {
        vector<int> indeg(numCourses,0);
        vector<int> mp[numCourses+1];
        queue<int> now_dedg_zero;
        for(int i = 0;i<prerequisites.size();i++)
        {
            int x = prerequisites[i][1];
            int y = prerequisites[i][0];
            indeg[y]++;
            mp[x].push_back(y);
        }
        int tot = 0;
        for(int i = 0;i<numCourses;i++) if(indeg[i] == 0) {now_dedg_zero.push(i);tot++;}
        while(!now_dedg_zero.empty())
        {
            int x = now_dedg_zero.front();
            now_dedg_zero.pop();
            for(int i = 0;i<mp[x].size();i++)
            {
                int y = mp[x][i];
                indeg[y] --;
                if(indeg[y] == 0) {now_dedg_zero.push(y);tot++;}
            }
        }
        if(tot == numCourses) return 1;
        return 0;
    }
};


int main()
{
    
    return 0;
}